### PR TITLE
SQADD/UQADD instructions

### PR DESCRIPTION
**Pull request by MS-DOS1999**

### CODE REVIEW

**Comment by gdkchan** [comment]
The implementation looks good, but imo it would be better to modify `EmitSaturatingNarrowOp` to also work without narrowing and use it for those instructions, to avoid introducing yet another method on `AInstEmitSimdHelper`.

For example, I recently modified the [EmitVectorShImmOp](https://github.com/Ryujinx/Ryujinx/blob/master/ChocolArm64/Instruction/AInstEmitSimdShift.cs#L309) method to also optionally add a round constant, by adding flags that controls what the method emits through a enum. I think something similar could be done there.

File: ChocolArm64/AOpCodeTable.cs [hunk 1]
```diff
@@ -373,6 +373,8 @@ static AOpCodeTable()
             SetA64("0x001110<<1xxxxx100000xxxxxxxxxx", AInstEmit.Smlal_V,       typeof(AOpCodeSimdReg));
             SetA64("0x001110<<1xxxxx101000xxxxxxxxxx", AInstEmit.Smlsl_V,       typeof(AOpCodeSimdReg));
             SetA64("0x001110<<1xxxxx110000xxxxxxxxxx", AInstEmit.Smull_V,       typeof(AOpCodeSimdReg));
+            SetA64("01011110xx1xxxxx000011xxxxxxxxxx", AInstEmit.Sqadd_S,       typeof(AOpCodeSimdReg));
+            SetA64("0x001110xx1xxxxx000011xxxxxxxxxx", AInstEmit.Sqadd_V,       typeof(AOpCodeSimdReg));
```
Comment (LDj3SNuD) [inline ChocolArm64/AOpCodeTable.cs]: `"0>001110<<1xxxxx000011xxxxxxxxxx"`

File: ChocolArm64/AOpCodeTable.cs [hunk 2]
```diff
@@ -421,6 +423,8 @@ static AOpCodeTable()
             SetA64("0x101110<<1xxxxx101011xxxxxxxxxx", AInstEmit.Uminp_V,       typeof(AOpCodeSimdReg));
             SetA64("0x001110000xxxxx001111xxxxxxxxxx", AInstEmit.Umov_S,        typeof(AOpCodeSimdIns));
             SetA64("0x101110<<1xxxxx110000xxxxxxxxxx", AInstEmit.Umull_V,       typeof(AOpCodeSimdReg));
+            SetA64("01111110xx1xxxxx000011xxxxxxxxxx", AInstEmit.Uqadd_S,       typeof(AOpCodeSimdReg));
+            SetA64("0x101110xx1xxxxx000011xxxxxxxxxx", AInstEmit.Uqadd_V,       typeof(AOpCodeSimdReg));
```
Comment (LDj3SNuD) [inline ChocolArm64/AOpCodeTable.cs]: `"0>101110<<1xxxxx000011xxxxxxxxxx"`

File: ChocolArm64/Instruction/AInstEmitSimdHelper.cs [hunk 3]
```diff
@@ -773,6 +773,115 @@ public static void EmitVectorPairwiseOp(AILEmitterCtx Context, Action Emit, bool
...
+        public static void EmitSaturatingOp(
+            AILEmitterCtx Context,
+            Action        Emit,
+            bool          SignedSrc,
+            bool          SignedDst,
+            bool          Scalar)
+        {
+            AOpCodeSimdReg Op = (AOpCodeSimdReg)Context.CurrOp;
+
+            int Elems = !Scalar ? 8 >> Op.Size : 1;
```
Comment (LDj3SNuD) [inline ChocolArm64/Instruction/AInstEmitSimdHelper.cs]:
```
int Bytes = Op.GetBitsCount() >> 3;
int Elems = !Scalar ? Bytes >> Op.Size : 1;
```

File: ChocolArm64/Instruction/AInstEmitSimdHelper.cs [hunk 4]
```diff
@@ -773,6 +773,115 @@ public static void EmitVectorPairwiseOp(AILEmitterCtx Context, Action Emit, bool
...
+
+            int ESize = 8 << Op.Size;
+
+            long TMaxValue = SignedDst ? (1 << (ESize - 1)) - 1 : (1L << ESize) - 1L;
```
Comment (LDj3SNuD) [inline ChocolArm64/Instruction/AInstEmitSimdHelper.cs]: `long TMaxValue = SignedDst ? (1 << (ESize - 1)) - 1 : (long)(ulong.MaxValue >> (64 - ESize));`

File: ChocolArm64/Instruction/AInstEmitSimdHelper.cs [hunk 5]
```diff
@@ -773,6 +773,115 @@ public static void EmitVectorPairwiseOp(AILEmitterCtx Context, Action Emit, bool
...
+
+            Context.EmitLdc_I8(0L);
+            Context.EmitSttmp();
+
+            for (int Index = 0; Index < Elems; Index++)
+            {
+                AILLabel LblLe    = new AILLabel();
+                AILLabel LblGeEnd = new AILLabel();
+
+                EmitVectorExtract(Context, Op.Rn, Index, Op.Size + 1, SignedSrc);
```
Comment (LDj3SNuD) [inline ChocolArm64/Instruction/AInstEmitSimdHelper.cs]: `EmitVectorExtract(Context, Op.Rn, Index, Op.Size, SignedSrc);`

File: ChocolArm64/Instruction/AInstEmitSimdHelper.cs [hunk 6]
```diff
@@ -773,6 +773,115 @@ public static void EmitVectorPairwiseOp(AILEmitterCtx Context, Action Emit, bool
...
+                EmitVectorExtract(Context, Op.Rm, Index, Op.Size + 1, SignedSrc);
```
Comment (LDj3SNuD) [inline ChocolArm64/Instruction/AInstEmitSimdHelper.cs]: `EmitVectorExtract(Context, Op.Rm, Index, Op.Size, SignedSrc);`

File: ChocolArm64/Instruction/AInstEmitSimdHelper.cs [hunk 7]
```diff
@@ -773,6 +773,115 @@ public static void EmitVectorPairwiseOp(AILEmitterCtx Context, Action Emit, bool
...
+                {
+                    EmitVectorZeroLower(Context, Op.Rd);
+                }
+
+                EmitVectorInsertTmp(Context, Index, Op.Size);
+            }
+
+            Context.EmitLdvectmp();
+            Context.EmitStvec(Op.Rd);
+
```
Comment (LDj3SNuD) [inline ChocolArm64/Instruction/AInstEmitSimdHelper.cs]:
```

if ((Op.RegisterSize == ARegisterSize.SIMD64) || Scalar)
{
    EmitVectorZeroUpper(Context, Op.Rd);
}

```

**Comment by MS-DOS1999** [comment]
@LDj3SNuD Thank you for the correction, I will change all that!

Just can you explain to me the meaning of "<" ">"
in "0> 001110 << 1xxxxx000011xxxxxxxxxx"
Gdkchan had explained to me a long time ago, but I forgot. Thank you !

**Comment by mailwl** [comment]
From source comment:

Note: < and > are used on special encodings.
The < means that we should never have ALL bits with the '<' set.
So, when the encoding has <<, it means that 00, 01, and 10 are valid,
but not 11. <<< is 000, 001, ..., 110 but NOT 111, and so on...
For >, the invalid value is zero. So, for >> 01, 10 and 11 are valid,
but 00 isn't.

**Comment by MS-DOS1999** [comment]
@gdkchan I need to make an enum too? Or just a boolean for Narrow / notNarrow?

**Comment by gdkchan** [comment]
IMO this method already has too many bools (SignedSrc, SignedDst, Scalar), and plus another one for Narrow, I think its way more readable if you use an enum for all those flags like I did on the inst I linked.

**Comment by MS-DOS1999** [comment]
`int Bytes = Op.GetBitsCount() >> 3;
int Elems = !Scalar ? Bytes >> Op.Size : 1;`

`long TMaxValue = SignedDst ? (1 << (ESize - 1)) - 1 : (long)(~0UL >> (64 - ESize));`

`EmitVectorExtract(Context, Op.Rn, Index, Op.Size, SignedSrc);`

`EmitVectorExtract(Context, Op.Rm, Index, Op.Size, SignedSrc);`

I tested this and several instructions do not work anymore with these changes, so I did not make these changes.

**Comment by gdkchan** [comment]
I think you should use `Op.Size` instead of `Op.Size + 1` when Narrow is false,

**Comment by gdkchan** [comment]
Closing in favor of #314.

**Pull request closed without merging**